File: bin/carve/gifs.py
#!/usr/bin/env python3
"""コマの PNG を並べて、動く GIF にする (標準ライブラリのみ)。

    python3 bin/gifs.py                # gallery/ の各工程から歩きの GIF を作る

止まった絵を並べても、コマ間の飛びや接地のずれは分からない。動かして初めて読める。
"""
import importlib.util
import os
import sys

HERE = os.path.dirname(os.path.abspath(__file__))
_s = importlib.util.spec_from_file_location("sheet", os.path.join(HERE, "sheet.py"))
sheet = importlib.util.module_from_spec(_s)
_s.loader.exec_module(sheet)

# 動きごとのコマの並び。歩きは接地→中間→接地(逆)→中間の 4 コマ
CYCLES = {
    "歩き": ("walk_0", "walk_1", "walk_2", "walk_3"),
    "振り": ("idle", "swing_0", "swing_1", "swing_2", "idle"),
    "跳ぶ": ("idle", "jump_0", "jump_1", "jump_2", "idle"),
}
DELAY = 14                                     # 1 コマの長さ (1/100 秒)


def _bits(codes, width_start, clear, end):
    """LZW の符号をビット列に詰める。辞書を伸ばさず、幅が上がる前に clear を打つ。

    圧縮しないぶん大きいが、符号化が 1 通りに決まるので壊れようがない。
    """
    out = bytearray()
    acc = acc_bits = 0
    width = width_start

    def push(code):
        nonlocal acc, acc_bits
        acc |= code << acc_bits
        acc_bits += width
        while acc_bits >= 8:
            out.append(acc & 0xFF)
            acc >>= 8
            acc_bits -= 8

    push(clear)
    since = 0
    for code in codes:
        push(code)
        since += 1
        if since >= 250:          # 復号側の辞書が幅を上げる前に区切り直す
            push(clear)
            since = 0
    push(end)
    if acc_bits:
        out.append(acc & 0xFF)
    return bytes(out)


def gif_of(width, height, frames, palette, delay=DELAY, transparent=0):
    """frames は「画素の番号の並び」の列。palette は (r,g,b) の並び (最大 256)。"""
    table = bytearray()
    for i in range(256):
        table.extend(palette[i] if i < len(palette) else (0, 0, 0))
    out = bytearray(b"GIF89a")
    out += bytes([width & 255, width >> 8, height & 255, height >> 8, 0xF7, 0, 0])
    out += table
    out += b"\x21\xFF\x0BNETSCAPE2.0\x03\x01\x00\x00\x00"   # 無限に繰り返す
    for pixels in frames:
        out += bytes([0x21, 0xF9, 4, 0b00001001, delay & 255, delay >> 8,
                      transparent, 0])
        out += bytes([0x2C, 0, 0, 0, 0, width & 255, width >> 8,
                      height & 255, height >> 8, 0])
        out += bytes([8])
        data = _bits(pixels, 9, 256, 257)
        for i in range(0, len(data), 255):
            chunk = data[i:i + 255]
            out += bytes([len(chunk)]) + chunk
        out += b"\x00"
    out += b"\x3B"
    return bytes(out)


def from_pngs(paths, shrink=1):
    """PNG を読んで、色を番号に振り直した GIF を作る。番号 0 は透明。

    shrink は工程の絵の拡大率。**等倍に戻してから**焼く — 拡大したまま焼くと、
    同じ絵なのに何十倍にも膨らむ。見るときの拡大は表示側の仕事。
    """
    read = []
    for path in paths:
        w, h, rows = sheet.read_png(path)
        if shrink > 1:
            rows = [[rows[y][x] for x in range(0, w, shrink)]
                    for y in range(0, h, shrink)]
            w, h = len(rows[0]), len(rows)
        read.append((w, h, rows))
    w = max(r[0] for r in read)
    h = max(r[1] for r in read)
    palette = [(0, 0, 0)]
    index = {}
    frames = []
    for pw, ph, rows in read:
        ox, oy = (w - pw) // 2, h - ph
        flat = []
        for y in range(h):
            for x in range(w):
                px = (0, 0, 0, 0)
                sy, sx = y - oy, x - ox
                if 0 <= sy < ph and 0 <= sx < pw:
                    px = rows[sy][sx]
                if px[3] == 0:
                    flat.append(0)
                    continue
                key = px[:3]
                if key not in index:
                    index[key] = len(palette)
                    palette.append(key)
                flat.append(index[key])
        frames.append(flat)
    return gif_of(w, h, frames, palette)


def main(argv):
    root = os.path.dirname(HERE)
    shrink = int(argv[0]) if argv else 6
    only = argv[1:]        # 工程を指定すると、その工程だけ焼く
    made = []
    for body in sorted(os.listdir(os.path.join(root, "gallery"))):
        home = os.path.join(root, "gallery", body)
        if not os.path.isdir(home):
            continue
        for stage in sorted(os.listdir(home)):
            if only and stage not in only:
                continue
            for view in sheet.VIEWS:
              for move, cycle in CYCLES.items():
                paths = [os.path.join(home, stage, f"{view}_{f}.png") for f in cycle]
                if not all(os.path.exists(p) for p in paths):
                    continue
                out = os.path.join(root, "gallery",
                                   f"{move}_{body}_{stage}_{view}.gif")
                with open(out, "wb") as handle:
                    handle.write(from_pngs(paths, shrink))
                made.append(os.path.relpath(out, root))
    print("\n".join(made))
    return 0


if __name__ == "__main__":
    sys.exit(main(sys.argv[1:]))

File: go/internal/carve/gifs.go
package carve

// コマの PNG を並べて動く GIF にする。
//
//	fge-go carve-gifs [縮小率] [工程名...]

import (
	"fmt"
	"os"
	"path/filepath"
	"strconv"
	"strings"
)

// GifCycle は動きごとのコマの並び。
type GifCycle struct {
	Name   string
	Frames []string
}

// bits は LZW の符号をビット列に詰める。辞書を伸ばさず、幅が上がる前に clear を打つ。
func bits(codes []int, widthStart, clear, end int) []byte {
	out := []byte{}
	acc, accBits := uint32(0), 0
	width := widthStart
	push := func(code int) {
		acc |= uint32(code) << accBits
		accBits += width
		for accBits >= 8 {
			out = append(out, byte(acc&0xFF))
			acc >>= 8
			accBits -= 8
		}
	}
	push(clear)
	since := 0
	for _, code := range codes {
		push(code)
		since++
		if since >= 250 {
			push(clear)
			since = 0
		}
	}
	push(end)
	if accBits != 0 {
		out = append(out, byte(acc&0xFF))
	}
	return out
}

// GifOf は画素の番号の並びから GIF を組む。
func GifOf(width, height int, frames [][]int, palette []RGB, delay, transparent int) []byte {
	table := make([]byte, 0, 768)
	for i := 0; i < 256; i++ {
		c := RGB{0, 0, 0}
		if i < len(palette) {
			c = palette[i]
		}
		table = append(table, byte(c[0]), byte(c[1]), byte(c[2]))
	}
	out := []byte("GIF89a")
	out = append(out, byte(width&255), byte(width>>8), byte(height&255), byte(height>>8), 0xF7, 0, 0)
	out = append(out, table...)
	out = append(out, []byte("\x21\xFF\x0BNETSCAPE2.0\x03\x01\x00\x00\x00")...)
	for _, pixels := range frames {
		out = append(out, 0x21, 0xF9, 4, 0b00001001, byte(delay&255), byte(delay>>8),
			byte(transparent), 0)
		out = append(out, 0x2C, 0, 0, 0, 0, byte(width&255), byte(width>>8),
			byte(height&255), byte(height>>8), 0)
		out = append(out, 8)
		data := bits(pixels, 9, 256, 257)
		for i := 0; i < len(data); i += 255 {
			end := i + 255
			if end > len(data) {
				end = len(data)
			}
			chunk := data[i:end]
			out = append(out, byte(len(chunk)))
			out = append(out, chunk...)
		}
		out = append(out, 0x00)
	}
	out = append(out, 0x3B)
	return out
}

// FromPNGs は PNG を読んで、色を番号に振り直した GIF を作る。番号 0 は透明。
func FromPNGs(paths []string, shrink int, delay int) ([]byte, error) {
	type sheetImage struct {
		w, h int
		rows [][]RGBA
	}
	var read []sheetImage
	for _, path := range paths {
		w, h, rows, err := ReadPNGSheet(path)
		if err != nil {
			return nil, err
		}
		if shrink > 1 {
			var small [][]RGBA
			for y := 0; y < h; y += shrink {
				var line []RGBA
				for x := 0; x < w; x += shrink {
					line = append(line, rows[y][x])
				}
				small = append(small, line)
			}
			rows = small
			w, h = len(rows[0]), len(rows)
		}
		read = append(read, sheetImage{w, h, rows})
	}
	w, h := 0, 0
	for _, r := range read {
		if r.w > w {
			w = r.w
		}
		if r.h > h {
			h = r.h
		}
	}
	palette := []RGB{{0, 0, 0}}
	index := map[RGB]int{}
	var frames [][]int
	for _, im := range read {
		ox, oy := (w-im.w)/2, h-im.h
		flat := make([]int, 0, w*h)
		for y := 0; y < h; y++ {
			for x := 0; x < w; x++ {
				px := RGBA{0, 0, 0, 0}
				sy, sx := y-oy, x-ox
				if 0 <= sy && sy < im.h && 0 <= sx && sx < im.w {
					px = im.rows[sy][sx]
				}
				if px[3] == 0 {
					flat = append(flat, 0)
					continue
				}
				key := RGB{px[0], px[1], px[2]}
				if _, ok := index[key]; !ok {
					index[key] = len(palette)
					palette = append(palette, key)
				}
				flat = append(flat, index[key])
			}
		}
		frames = append(frames, flat)
	}
	return GifOf(w, h, frames, palette, delay, 0), nil
}

// RunGifs は carve-gifs の入口。root は assets / gallery を置く根。
func RunGifs(out *strings.Builder, root string, args []string, rules *Rules) (int, error) {
	shrink := 6
	if len(args) > 0 {
		n, err := strconv.Atoi(args[0])
		if err != nil {
			return 1, err
		}
		shrink = n
	}
	var only []string
	if len(args) > 1 {
		only = args[1:]
	}
	var made []string
	bodies, err := sortedDir(filepath.Join(root, "gallery"))
	if err != nil {
		return 1, err
	}
	for _, body := range bodies {
		home := filepath.Join(root, "gallery", body)
		if !isDir(home) {
			continue
		}
		stages, err := sortedDir(home)
		if err != nil {
			return 1, err
		}
		for _, stage := range stages {
			if len(only) > 0 && !contains(only, stage) {
				continue
			}
			for _, view := range SheetViews {
				for _, cycle := range rules.GifCycles {
					var paths []string
					ok := true
					for _, f := range cycle.Frames {
						p := filepath.Join(home, stage, fmt.Sprintf("%s_%s.png", view, f))
						if _, err := os.Stat(p); err != nil {
							ok = false
						}
						paths = append(paths, p)
					}
					if !ok {
						continue
					}
					outPath := filepath.Join(root, "gallery",
						fmt.Sprintf("%s_%s_%s_%s.gif", cycle.Name, body, stage, view))
					data, err := FromPNGs(paths, shrink, rules.GifDelay)
					if err != nil {
						return 1, err
					}
					if err := os.WriteFile(outPath, data, 0o644); err != nil {
						return 1, err
					}
					rel, _ := filepath.Rel(root, outPath)
					made = append(made, rel)
				}
			}
		}
	}
	fmt.Fprintln(out, strings.Join(made, "\n"))
	return 0, nil
}

func contains(list []string, v string) bool {
	for _, s := range list {
		if s == v {
			return true
		}
	}
	return false
}
